### ISSUE TITLE
Implementing Softmax Cross-entropy with SameDiff

### ISSUE THREAD

**Issue by nadheesh** [deeplearning4j/deeplearning4j#6606]
I'm trying to implement Softmax Cross-entropy with SameDiff. I found that SameDiff offers sd.softmaxCrossEntropyWithLogits(), but it does not seem to implement gradients. I got this error `"java.lang.UnsupportedOperationException: Please extend DynamicCustomOp.doDiff to support SameDiff backprop operations.`".

Therefore, I tried implementing Softmax cross entropy myself using SDVariables. This is what I have done so far based on the typical cross entropy loss funtion

![image](https://user-images.githubusercontent.com/12269120/47134473-7c927580-d2ca-11e8-8b58-f726fb793790.png)

```
public static SDVariable softmaxCrossentropyWithLogits(SDVariable logits, SDVariable labels, SameDiff sd){
        SDVariable logq = sd.log(sd.softmax(logits)); //didn't use logSoftmax because it seems to have zero gradient issue #6605
        return sd.neg(logq.mul(labels).sum(1)); // implements the equation. labels are one-hot encoded
    }
```

Even though, this works well with certain problems, sometimes I encounter infinity as the loss (since logq becomes infinity when sd.softmax(logits) == 0) resulting in invalid gradients. I minimize the number of such cases by avoiding log computation of unnecessary cases as show below. 

```
public static SDVariable softmaxCrossentropyWithLogits(SDVariable logits, SDVariable labels, SameDiff sd){
        SDVariableq = sd.softmax(logits); 
        return sd.neg(sd.log(logq.mul(labels).sum(1))); // labels are one-hot encoded. log is taken after multiplying with one-hot vectors and taking the sum of resulting vectors. 
    }
```

I found that tensorflow use a transformed version of cross-entropy [1](https://stats.stackexchange.com/questions/348036/difference-between-mathematical-and-tensorflow-implementation-of-softmax-crossen) as shown in the equation below.

![image](https://user-images.githubusercontent.com/12269120/47134884-ed865d00-d2cb-11e8-83ae-86778017f062.png)

I'm planing to implement this, yet I'm stuck at computing the z_i - max(z) based on batches, where we have perform a  `sub` operator between variables with shapes [batch_size, num_classes] and [batch_size]. 

My questions are,

1. what are the optimal ways that we can use to implement Softmax cross-entropy loss using SameDiff?
2. If we are to follow, tensorflow implementation, what is the best way of performing above substitution with batches?

Thanks!!!

**Comment by AlexDBlack** [comment]
Briefly: unfortunately yes, we do need the loss function gradients implemented.
There's an issue open for that here: https://github.com/deeplearning4j/deeplearning4j/issues/6517

The implementation will happen at the c++ level, but the c++ team is busy with other tasks at present.

As for implementing it more directly, the DL4J implementation may be worth a look, for one approach:
https://github.com/deeplearning4j/deeplearning4j/blob/master/nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/lossfunctions/impl/LossMCXENT.java
That's different to what TF does, but we use clipping to avoid numerical underflow instead.

> I'm planing to implement this, yet I'm stuck at computing the z_i - max(z) based on batches, where we have perform a sub operator between variables with shapes [batch_size, num_classes] and [batch_size].

Assuming you mean m = max(z) with shape [batch_size], then you can use 'keep_dims=true' to get a [batch_size,1] result, which can be auto-broadcast with the [batch_size, num_classes] - i.e., z.sub(z.max(true,1)) is "subtract max (vector) from each column"
https://github.com/deeplearning4j/deeplearning4j/blob/06ab49e6812fbb8d4ad81833977ff807840e6eab/nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/SameDiff.java#L6391-L6406

SDVariable.max(boolean, int...) is the same thing (no javadoc there though)

**Comment by nadheesh** [comment]
Thanks @AlexDBlack . 

If I want to use `sd.clipByValue`, how can I only use clip by max, disabling the clip by min. I tried using Double.MIN as the `clipValueMin`, but it did not work.

**Comment by AlexDBlack** [comment]
Sorry, I missed your last comment here.
Clipping with Double.MIN_VALUE will clip just above zero.

For the one-sided clipping, just use .scalarMax(SDVariable, Number) or scalarMin(SDVariable, Number)
